### DISCUSSION TITLE
Use openssl.cnf with python (urllib)

### DISCUSSION THREAD

**Discussion by baentsch** [openssl/openssl#22359] [Q&A]
Advance warning: I'm aware this may be better asked to the Python community -- but I don't know of the best place to do that, so any pointers welcome ([SO question here for reference](https://stackoverflow.com/questions/77278628/how-to-set-tls1-3-groups-to-use-when-using-pythons-urllib-request-api)).

As the issue is one of OpenSSL (integrated to Python) let me ask it regardless hoping some "pythonite" sees it:

Activating a (new) provider via "openssl.cnf" finally works in Python since/as openssl3 support has been enabled there.
However, it seems the file is used only once at startup and not consulted again when doing a new `urllib.request`. Is this known/intentional behaviour?

Background: Python/urllib does not provide an API to set (TLS1.3) default groups, so I try to achieve this (changing the default groups per connection) by (changing an) environment variable in "openssl.cnf", e.g., as is [documented here](https://www.openssl.org/blog/blog/2022/10/21/tls-groups-configuration). This works fine outside of Python but not within.

So the question may be: How can one trigger re-evaluation of "openssl.cnf" (using the then-current environment variables) within a running Python program (to ultimately achieve the same as calling [SSL_set1_groups](https://www.openssl.org/docs/man3.1/man3/SSL_set1_groups.html]))?

**Comment by t8m** [comment]
I am afraid there is no way to reload the default configuration file in this use case. Or rather, there are ways to do that using C APIs such as CONF_modules_load_file() but I do not think Python provides them.

**Reply by baentsch** [reply]
Thanks for the response and suggestion... But nope: Not a single reference to "CONF_" in the whole python ssl API. But well, I anyway only need this for testing (all KEMs in a certain provider). Will revert to spawning different processes then. So much about efficiency... but then again, this is Python :-)
